File: src/com/castlefrog/agl/domains/yahtzee/YahtzeeState.java
package com.castlefrog.agl.domains.yahtzee;

import com.castlefrog.agl.State;

public final class YahtzeeState implements State<YahtzeeState> {
    public static final int N_DICE = 5;
    public static final int N_VALUES = 6;
    public static final int N_SCORES = YahtzeeScoreCategory.values().length;

    private byte[] diceValues_;
    private byte rolls_;
    private int[] scores_;

    public YahtzeeState(byte[] diceValues, int rolls, int[] scores) {
        diceValues_ = diceValues;
        rolls_ = (byte) rolls;
        scores_ = scores;
    }

    public YahtzeeState copy() {
        return this;
    }

    public int getDiceValue(int index) {
        return diceValues_[index];
    }

    public byte[] getDiceValues() {
        byte[] diceValues = new byte[N_VALUES];
        for (int i = 0; i < N_VALUES; i++) {
            diceValues[i] = diceValues_[i];
        }
        return diceValues;
    }

    public int getRolls() {
        return rolls_;
    }

    public int[] getScores() {
        int[] scores = new int[N_SCORES];
        for (int i = 0; i < N_SCORES; i++) {
            scores[i] = scores_[i];
        }
        return scores;
    }

    public int getScore(int index) {
        return scores_[index];
    }

    public int getAgentTurn() {
        return 0;
    }

    @Override
    public int hashCode() {
        int code = 7 + rolls_;
        for (int i = 0; i < N_VALUES; i++) {
            code = 11 * code + diceValues_[i];
        }
        for (int i = 0; i < N_SCORES; i++) {
            code = 11 * code + scores_[i];
        }
        return code;
    }

    @Override
    public boolean equals(Object object) {
        if (!(object instanceof YahtzeeState)) {
            return false;
        }
        YahtzeeState state = (YahtzeeState) object;
        for (int i = 0; i < N_VALUES; i++) {
            if (diceValues_[i] != state.getDiceValue(i)) {
                return false;
            }
        }
        if (rolls_ != state.getRolls()) {
            return false;
        }
        for (int i = 0; i < N_SCORES; i++) {
            if (scores_[i] != state.getScore(i)) {
                return false;
            }
        }
        return true;
    }

    @Override
    public String toString() {
        StringBuilder output = new StringBuilder();
        YahtzeeScoreCategory[] scoreCategories = YahtzeeScoreCategory.values();
        for (int i = 0; i < N_SCORES; i++) {
            if (scores_[i] != -1) {
                output.append(scoreCategories[i].name() + ": " + scores_[i] + "\n");
            }
        }
        output.append("Rolls: " + rolls_ + "\n");
        output.append("Dice: [ ");
        for (int i = 0; i < diceValues_.length; i++) {
            output.append(diceValues_[i] + " ");
        }
        output.append("]\n");
        for (int i = 0; i < N_SCORES; i++) {
            if (scores_[i] == -1) {
                output.append(scoreCategories[i].name() + " ");
            }
        }
        return output.toString();
    }
}

File: src/main/kotlin/com/castlefrog/agl/domains/yahtzee/YahtzeeState.kt
package com.castlefrog.agl.domains.yahtzee

import com.castlefrog.agl.State

data class YahtzeeState(
    val diceValues: ByteArray,
    var nRolls: Byte = 1,
    val scores: IntArray = IntArray(N_SCORES) { -1 }
) : State<YahtzeeState> {

    companion object {
        const val N_DICE = 5
        const val N_VALUES = 6
        val N_SCORES = YahtzeeScoreCategory.values().size
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other?.javaClass != javaClass) return false
        other as YahtzeeState
        return nRolls == other.nRolls &&
                diceValues.contentEquals(other.diceValues) &&
                scores.contentEquals(other.scores)
    }

    override fun hashCode(): Int {
        return (nRolls.toInt() * 17 + diceValues.contentHashCode()) * 19 + scores.contentHashCode()
    }

    override fun copy(): YahtzeeState {
        return copy(diceValues = diceValues, nRolls = nRolls, scores = scores)
    }

    override fun toString(): String {
        val output = StringBuilder()
        output.append(nRolls).append(" - [ ")
        for (diceValue in diceValues) {
            output.append(diceValue).append(" ")
        }
        output.append("]\n")
        val scoreCategories = YahtzeeScoreCategory.values()
        for (i in 0 until N_SCORES) {
            output.append(scoreCategories[i].name).append(": ")
            output.append(if (scores[i] != -1) scores[i] else "-")
            if (i != N_SCORES - 1) {
                output.append("\n")
            }
        }
        return output.toString()
    }
}
